pkg/logql/series_extractor_test.go

```diff
@@ -0,0 +1,146 @@
+package logql
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/loki/pkg/iter"
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+func Test_seriesIterator_Peek(t *testing.T) {
+	type expectation struct {
+		ok     bool
+		sample Sample
+	}
+	for _, test := range []struct {
+		name         string
+		it           SeriesIterator
+		expectations []expectation
+	}{
+		{
+			"count",
+			newSeriesIterator(iter.NewStreamIterator(newStream(5, identity, `{app="foo"}`)), extractCount),
+			[]expectation{
+				{true, Sample{Labels: `{app="foo"}`, TimestampNano: 0, Value: 1}},
+				{true, Sample{Labels: `{app="foo"}`, TimestampNano: time.Unix(1, 0).UnixNano(), Value: 1}},
+				{true, Sample{Labels: `{app="foo"}`, TimestampNano: time.Unix(2, 0).UnixNano(), Value: 1}},
+				{true, Sample{Labels: `{app="foo"}`, TimestampNano: time.Unix(3, 0).UnixNano(), Value: 1}},
+				{true, Sample{Labels: `{app="foo"}`, TimestampNano: time.Unix(4, 0).UnixNano(), Value: 1}},
+				{false, Sample{}},
+			},
+		},
+		{
+			"bytes empty",
+			newSeriesIterator(
+				iter.NewStreamIterator(newStream(3,
+					func(i int64) logproto.Entry {
+						return logproto.Entry{
+							Timestamp: time.Unix(i, 0),
+						}
+					},
+					`{app="foo"}`)),
+				extractBytes,
+			),
+			[]expectation{
+				{true, Sample{Labels: `{app="foo"}`, TimestampNano: 0, Value: 0}},
+				{true, Sample{Labels: `{app="foo"}`, TimestampNano: time.Unix(1, 0).UnixNano(), Value: 0}},
+				{true, Sample{Labels: `{app="foo"}`, TimestampNano: time.Unix(2, 0).UnixNano(), Value: 0}},
+				{false, Sample{}},
+			},
+		},
+		{
+			"bytes",
+			newSeriesIterator(
+				iter.NewStreamIterator(newStream(3,
+					func(i int64) logproto.Entry {
+						return logproto.Entry{
+							Timestamp: time.Unix(i, 0),
+							Line:      "foo",
+						}
+					},
+					`{app="foo"}`)),
+				extractBytes,
+			),
+			[]expectation{
+				{true, Sample{Labels: `{app="foo"}`, TimestampNano: 0, Value: 3}},
+				{true, Sample{Labels: `{app="foo"}`, TimestampNano: time.Unix(1, 0).UnixNano(), Value: 3}},
+				{true, Sample{Labels: `{app="foo"}`, TimestampNano: time.Unix(2, 0).UnixNano(), Value: 3}},
+				{false, Sample{}},
+			},
+		},
+		{
+			"bytes backward",
+			newSeriesIterator(
+				iter.NewStreamsIterator(context.Background(),
+					[]logproto.Stream{
+						newStream(3,
+							func(i int64) logproto.Entry {
+								return logproto.Entry{
+									Timestamp: time.Unix(i, 0),
+									Line:      "foo",
+								}
+							},
+							`{app="foo"}`), newStream(3,
+							func(i int64) logproto.Entry {
+								return logproto.Entry{
+									Timestamp: time.Unix(i, 0),
+									Line:      "barr",
+								}
+							},
+							`{app="barr"}`),
+					},
+					logproto.BACKWARD,
+				),
+				extractBytes,
+			),
+			[]expectation{
```

Comment:
Shouldn't these alternate label sets (foo -> bar -> foo, etc) and shouldn't the timestamps be in desc order due to `logproto.BACKWARD`?

Reply:
Backward doesn't change anything for metrics. That's what I'm showcasing here, and it should not.